File: 프로그래머스/3/42628. 이중우선순위큐/이중우선순위큐.java
import java.util.*;
import java.io.*;


class Solution {
    public static class Node{
        int num;
        int index;
        public Node(int num,int index){
            this.num=num;
            this.index=index;
        }
        
    }
    
    public int[] solution(String[] operations) {
        PriorityQueue<Node> pq1=new PriorityQueue<>((o1,o2)->Integer.compare(o1.num,o2.num));
        PriorityQueue<Node> pq2=new PriorityQueue<>((o1,o2)->Integer.compare(o2.num,o1.num));
        int[] deleted=new int[operations.length];
        
        for(int i = 0;i<operations.length;i++){
            StringTokenizer st=new StringTokenizer(operations[i]);
            String command=st.nextToken();
            int num=Integer.parseInt(st.nextToken());
            if(command.equals("I")){
                pq1.offer(new Node(num,i));
                pq2.offer(new Node(num,i));
                deleted[i]=1;
            }
            else{
                if(num==1){
                    // 최댓값 삭제
                    while(!pq1.isEmpty() && deleted[pq1.peek().index]!=1){
                        pq1.poll();
                    }
                    while(!pq2.isEmpty()&&deleted[pq2.peek().index]!=1){
                        pq2.poll();
                    }
                    if(!pq2.isEmpty()){
                        deleted[pq2.peek().index]=0;
                        pq2.poll();
                    }
                }
                else{
                    // 최솟값 삭제
                    while(!pq1.isEmpty() && deleted[pq1.peek().index]!=1){
                        pq1.poll();
                    }
                    while(!pq2.isEmpty()&&deleted[pq2.peek().index]!=1){
                        pq2.poll();
                    }
                    if(!pq1.isEmpty()){
                        deleted[pq1.peek().index]=0;
                        pq1.poll();
                    }
                    
                }
                
                
            }
        }
        int[] answer=new int[2];
        while(!pq1.isEmpty() && deleted[pq1.peek().index]!=1){
            System.out.println(pq1.peek().num);
            pq1.poll();
        }
        while(!pq2.isEmpty()&&deleted[pq2.peek().index]!=1){
            System.out.println(pq2.peek().num);
            pq2.poll();
        }
        if(!pq2.isEmpty()){
            
        }
        if(!pq1.isEmpty()){
            
        }
        
        if(!pq1.isEmpty()){
            answer[1]=pq1.peek().num;
        }
        if(!pq2.isEmpty()){
            answer[0]=pq2.peek().num;
        }
        return answer;
        
    }
}

File: 프로그래머스/3/42628. 이중우선순위큐/이중우선순위큐.py
import heapq

def solution(operations):
    
    deleted=[0]*len(operations)
    min_hq=[]
    max_hq=[]
    for i in range(len(operations)):
        command,num=operations[i].split()
        num=int(num)
        if command=="I":
            heapq.heappush(min_hq,(num,i))
            heapq.heappush(max_hq,(-num,i))
            deleted[i]=1
        elif command=="D":
            
            
            # 최댓값 삭제
            if num==1:
                while min_hq and deleted[min_hq[0][1]]!=1:
                    heapq.heappop(min_hq)
                while max_hq and deleted[max_hq[0][1]]!=1:
                    heapq.heappop(max_hq)
                
                
                if max_hq:
                    deleted[max_hq[0][1]]=0
                    heapq.heappop(max_hq)
                
                
            # 최솟값 삭제
            else:
                while min_hq and deleted[min_hq[0][1]]!=1:
                    heapq.heappop(min_hq)
                while max_hq and deleted[max_hq[0][1]]!=1:
                    heapq.heappop(max_hq)
                if min_hq:
                    deleted[min_hq[0][1]]=0
                    heapq.heappop(min_hq)
                
    result=[]
    while min_hq and deleted[min_hq[0][1]]!=1:
        heapq.heappop(min_hq)
    while max_hq and deleted[max_hq[0][1]]!=1:
        heapq.heappop(max_hq)
    
    if max_hq and deleted[max_hq[0][1]]==1:
        result.append(-heapq.heappop(max_hq)[0])
    else:
        result.append(0)
    
    if min_hq and deleted[min_hq[0][1]]==1:
        result.append(heapq.heappop(min_hq)[0])
    else:
        result.append(0)
    return result
                
            
    